File: 5 DFS/Lint_595.java
// https://www.lintcode.com/problem/binary-tree-longest-consecutive-sequence/description?_from=ladder&&fromId=1

// M1 top-down w/ global var (more like mix, top-down + bottom-up)
public class Solution {
    /**
     * @param root: the root of binary tree
     * @return: the length of the longest consecutive sequence path
     */
    int maxLen = 0;
    public int longestConsecutive(TreeNode root) {
        helper(root, null, 0);
        
        return maxLen;
    }
    
    private void helper(TreeNode root, TreeNode parent, int len) {
        if (root == null)
            return;
        
        len = (parent != null && parent.val + 1 == root.val) ?
              len + 1 :
              1;
        maxLen = Math.max(len, maxLen);
        
        helper(root.left, root, len);
        helper(root.right, root, len);
        
    }
    
// M1 - 2: top-down without global var (more like mix, top-down + bottom-up)

public int longestConsecutive(TreeNode root) {
    return dfs(root, null, 0);
}

private int dfs(TreeNode p, TreeNode parent, int length) {
    if (p == null) return length;
    length = (parent != null && p.val == parent.val + 1) ? length + 1 : 1;
    return Math.max(length, Math.max(dfs(p.left, p, length),
                                     dfs(p.right, p, length)));
}

// M2, bottom-up
private int maxLength = 0;
public int longestConsecutive(TreeNode root) {
    dfs(root);
    return maxLength;
}

private int dfs(TreeNode p) {
    if (p == null) return 0;
    int L = dfs(p.left) + 1;
    int R = dfs(p.right) + 1;
    if (p.left != null && p.val + 1 != p.left.val) {
        L = 1;
    }
    if (p.right != null && p.val + 1 != p.right.val) {
        R = 1;
    }
    int length = Math.max(L, R);
    maxLength = Math.max(maxLength, length);
    return length;
}

File: 5 DFS_py/Lint_595.py
class Solution:
    """
    @param root: the root of binary tree
    @return: the length of the longest consecutive sequence path
    """
    def longestConsecutive(self, root):
        _, max_len = self.helper(root)
        return max_len
    
    #return curt_len, max_len
    def helper(self, root):
        if root is None:
            return 0, 0
            
        l_curt_len, l_max_len = self.helper(root.left)
        r_curt_len, r_max_len = self.helper(root.right)
        
        c_curt_len = 1
        c_max_len = 0
        
        if root.left and root.val + 1 == root.left.val:
            c_curt_len = max(c_curt_len, l_curt_len + 1)
        
        if root.right and root.val + 1 == root.right.val:
            c_curt_len = max(c_curt_len, r_curt_len + 1)
        
        c_max_len = max(c_curt_len, l_max_len, r_max_len)
        
        return c_curt_len, c_max_len
